types/time.go

```diff
@@ -528,36 +528,78 @@ func TimestampDiff(unit string, t1 Time, t2 Time) int64 {
 	return timestampDiff(unit, t1.Time, t2.Time)
 }
 
-// ParseDateFormat parses a formatted date string and returns separated components.
-func ParseDateFormat(format string) []string {
-	format = strings.TrimSpace(format)
+type parseFormatHelper struct {
+	format string
+	idx    int
+	hasT   bool
+}
 
-	start := 0
-	var seps = make([]string, 0)
-	for i := 0; i < len(format); i++ {
-		// Date format must start and end with number.
-		if i == 0 || i == len(format)-1 {
-			if !unicode.IsNumber(rune(format[i])) {
-				return nil
-			}
+func (h *parseFormatHelper) valid() bool {
+	return h.idx >= 0 && h.idx < len(h.format)
+}
 
-			continue
-		}
+func (h *parseFormatHelper) parse() []string {
+	var seps []string
+	for h.valid() {
+		str := h.parseNumber()
+		h.parseSeperate()
+		seps = append(seps, str)
+	}
+	if h.idx < 0 {
+		return nil
+	}
+	return seps
+}
 
-		// Separator is a single none-number char.
-		if !unicode.IsNumber(rune(format[i])) {
-			if !unicode.IsNumber(rune(format[i-1])) {
-				return nil
-			}
+func (h *parseFormatHelper) parseNumber() string {
+	start := h.idx
+	for h.valid() && unicode.IsNumber(rune(h.format[h.idx])) {
+		h.idx++
+	}
+	if start != h.idx {
+		return h.format[start:h.idx]
+	}
+	h.idx = -1
+	return ""
+}
 
-			seps = append(seps, format[start:i])
-			start = i + 1
+func (h *parseFormatHelper) parseSeperate() {
+	if !h.valid() {
+		return
+	}
+	// Separator is a single none-number char or whitespaces.
+	switch h.format[h.idx] {
+	case ':', '-', '^', '+', '/', '@', '~', '.':
```

Comment:
what about '(' and so on? you can exec below in MySQL:

```
mysql> select date('2011-11-11(10:10:10');
+-----------------------------+
| date('2011-11-11(10:10:10') |
+-----------------------------+
| 2011-11-11                  |
+-----------------------------+
1 row in set (0.00 sec)
```

Reply:
Now I begin to doubt whether it worth the  change...